src/cursor/abstract_cursor.ts

```diff
@@ -0,0 +1,697 @@
+  comment?: Document | string;
+  tailable?: boolean;
+  awaitData?: boolean;
+  noCursorTimeout?: boolean;
+}
+
+/** @internal */
+export type InternalAbstractCursorOptions = Omit<AbstractCursorOptions, 'readPreference'> & {
+  // resolved
+  readPreference: ReadPreference;
+
+  // cursor flags, some are deprecated
+  oplogReplay?: boolean;
+  exhaust?: boolean;
+  partial?: boolean;
+};
+
+/** @internal */
+export abstract class AbstractCursor extends EventEmitter {
+  [kId]?: Long;
+  [kSession]?: ClientSession;
+  [kServer]?: Server;
+  [kNamespace]: MongoDBNamespace;
+  [kDocuments]: Document[];
+  [kTopology]: Topology;
+  [kTransform]?: (doc: Document) => Document;
+  [kClosed]: boolean;
+  [kKilled]: boolean;
+  [kOptions]: InternalAbstractCursorOptions;
+
+  /** @event */
+  static readonly CLOSE = 'close' as const;
+
+  constructor(
+    topology: Topology,
+    namespace: MongoDBNamespace,
+    options: AbstractCursorOptions = {}
+  ) {
+    super();
+
+    this[kTopology] = topology;
+    this[kNamespace] = namespace;
+    this[kDocuments] = []; // TODO: https://github.com/microsoft/TypeScript/issues/36230
+    this[kClosed] = false;
+    this[kKilled] = false;
+    this[kOptions] = {
+      readPreference:
+        options.readPreference && options.readPreference instanceof ReadPreference
+          ? options.readPreference
+          : ReadPreference.primary,
+      ...pluckBSONSerializeOptions(options)
+    };
+
+    if (typeof options.batchSize === 'number') {
+      this[kOptions].batchSize = options.batchSize;
+    }
+
+    if (typeof options.comment !== 'undefined') {
+      this[kOptions].comment = options.comment;
+    }
+
+    if (typeof options.maxTimeMS === 'number') {
+      this[kOptions].maxTimeMS = options.maxTimeMS;
+    }
+
+    if (options.session instanceof ClientSession) {
+      this[kSession] = options.session;
+    }
+  }
+
+  get id(): Long | undefined {
+    return this[kId];
+  }
+
+  get topology(): Topology {
+    return this[kTopology];
+  }
+
+  get server(): Server | undefined {
+    return this[kServer];
+  }
+
+  get namespace(): MongoDBNamespace {
+    return this[kNamespace];
+  }
+
+  get readPreference(): ReadPreference {
+    return this[kOptions].readPreference;
+  }
+
+  get session(): ClientSession | undefined {
+    return this[kSession];
+  }
+
+  /** @internal */
+  get cursorOptions(): InternalAbstractCursorOptions {
+    return this[kOptions];
+  }
+
+  get closed(): boolean {
+    return this[kClosed];
+  }
+
+  get killed(): boolean {
+    return this[kKilled];
+  }
+
+  // NOTE: should we remove these? They are currently needed by a number of tests
+  isClosed(): boolean {
+    return this.closed;
+  }
+
+  /** Returns current buffered documents length */
+  bufferedCount(): number {
+    return this[kDocuments].length;
+  }
+
+  /** Returns current buffered documents */
+  readBufferedDocuments(number: number): Document[] {
+    return this[kDocuments].splice(0, number);
+  }
```

Comment:
Good point, though I think it might be something more like:
```
return this[kDocuments].splice(0, Math.min(number ?? this.bufferedCount()));
```

FWIW I really despise the "bufferedX" methods, and think we should defer to some solution like providing a `nextBatch` method.

Reply:
The number param still needs to be made optional, but also looking closer at splice's behavior if TS will allow it you can pass the potentially undefined number directly to splice and it will default to grabbing [0, length of the array]. 

This is moot if you just use the undefined number directly but if not was it intentional to change to using `this[kDocuments].length` as opposed to `this.bufferedCount()`?

Also the documentation still needs a note here that it removes from the buffer.

Or if all this is moot for the pending `nextBatch()` solution, is there a ticket floating around for that?

Reply:
Fixed this up. We don't currently have a ticket for `nextBatch` explicitly, but [NODE-2860](https://jira.mongodb.org/browse/NODE-2860) suggests it.